File: src/main/java/dev/kensa/util/SourceCodeIndex.java
package dev.kensa.util;

import java.io.IOException;
import java.nio.file.*;
import java.nio.file.attribute.BasicFileAttributes;

import static java.lang.System.getProperty;

public class SourceCodeIndex {

    private static final SourceCodeIndex INSTANCE = new SourceCodeIndex();

    public static Path locate(Class<?> clazz) {
        return INSTANCE.doLocate(clazz);
    }

    private final Path workingDirectory = Paths.get(getProperty("user.dir"));

    private SourceCodeIndex() {
    }

    private Path doLocate(Class<?> clazz) {
        String name = clazz.getName();

        int s = name.indexOf("$");
        if(s > 0) {
            name = name.substring(0, s);
        }
        Walker walker = new Walker(name);
        try {
            Files.walkFileTree(workingDirectory, walker);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        return walker.result();
    }

    private static class Walker implements FileVisitor<Path> {
        private final PathMatcher pathMatcher;
        private Path result;

        Walker(String name) {
            pathMatcher = FileSystems.getDefault().getPathMatcher(String.format("glob:**/%s.java", name.replaceAll("\\.", "/")));
        }

        @Override
        public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) {
            return FileVisitResult.CONTINUE;
        }

        @Override
        public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
            if (pathMatcher.matches(file)) {
                result = file;
            }

            return FileVisitResult.CONTINUE;
        }

        @Override
        public FileVisitResult visitFileFailed(Path file, IOException exc) {
            return FileVisitResult.CONTINUE;
        }

        @Override
        public FileVisitResult postVisitDirectory(Path dir, IOException exc) {
            return FileVisitResult.CONTINUE;
        }

        Path result() {
            return result;
        }
    }
}

File: src/main/java/dev/kensa/util/SourceCodeIndex.kt
package dev.kensa.util

import dev.kensa.KensaException
import java.io.IOException
import java.nio.file.*
import java.nio.file.FileVisitResult.CONTINUE
import java.nio.file.attribute.BasicFileAttributes
import kotlin.reflect.KClass
import kotlin.reflect.jvm.jvmName

object SourceCodeIndex {
    private val workingDirectory = Path.of(System.getProperty("user.dir"))

    private val KClass<*>.sourceExtension get() = if (isKotlinClass) "kt" else "java"

    fun locate(kClass: KClass<*>): Path =
        kClass.jvmName.substringBefore("$").let { name ->
            Walker(name, kClass.sourceExtension).run {
                Files.walkFileTree(workingDirectory, this)
                result ?: throw KensaException("Could not locate source file for [$name]")
            }
        }

    private class Walker(name: String, sourceExtension: String) : FileVisitor<Path?> {
        private val pathMatcher: PathMatcher = FileSystems.getDefault().getPathMatcher("glob:**/${name.replace("\\.".toRegex(), "/")}.$sourceExtension")
        var result: Path? = null

        override fun preVisitDirectory(dir: Path?, attrs: BasicFileAttributes) = CONTINUE

        override fun visitFile(file: Path?, attrs: BasicFileAttributes) = CONTINUE.apply { if (pathMatcher.matches(file)) result = file }

        override fun visitFileFailed(file: Path?, exc: IOException?) = CONTINUE

        override fun postVisitDirectory(dir: Path?, exc: IOException?) = CONTINUE
    }
}